### PR TITLE
Add turretinfobox support

### PR DESCRIPTION
**Pull request by sharpy66**
This pull request primarily introduces support for automatically adding and updating tooltip (pop-up) turret statistics within ship infoboxes (see discord). It also includes some miscellaneous fixes for issues you may or may not have.

**Notes:**
-
* Unless otherwise stated, changes have been tested to a reasonable extent locally
* C# is not a language I have much experience in, so there are likely issues I am not aware of

**Per-file summary:**
-
**ApiManager.cs**
* Corrected API endpoint from `ships-turret` to `ships-turrets`
* Fixed an error I had with Json de-serialization (may be platform-specific)

**TurretData.cs**
* Fixed an error I had with Json de-serialization (may be platform-specific)
* Updated types and made turret data fields nullable to gracefully handle missing or optional API data
* Added a `TurretTypeEnum` for the new `Beam` turret type

**TurretUpdater.cs**
* Implemented a method to automatically update cargo table data for turrets
* Escaped turret tables to prevent the Regex engine trying (and failing) to interpret special characters within them
* Added support for new `Beam` type turrets (not tested)

**WikiParser.cs**
* Updated ShipInfobox extraction Regex to work with nested templates (may require further testing for edge-cases)
* Allow up to 7 turret stat tables to accommodate the new `Beam` type

**ExtractTurretTablesTests.cs**
* Update test to reflect increased table limit

**ShipUpdater.cs**
* Implemented logic to wrap turrets with `Tooltip` and `TurretInfobox` templates

### CODE REVIEW

File: ketchupbot-framework/API/ApiManager.cs [hunk 1]
```diff
@@ -82,6 +82,6 @@ public async Task<Dictionary<string, Dictionary<string, string>>> GetShipsData()
 
         dynamic? jsonResponse = JsonConvert.DeserializeObject<dynamic>(stringResponse);
 
-        return JsonConvert.DeserializeObject<Dictionary<string, TurretData>>(jsonResponse?.serializedTurrets);
+        return JsonConvert.DeserializeObject<Dictionary<string, TurretData>>(jsonResponse?.serializedTurrets.ToString());
```
Comment (smallketchup82) [inline ketchupbot-framework/API/ApiManager.cs]: This is redundant

File: ketchupbot-framework/TurretUpdater.cs [hunk 2]
```diff
@@ -47,15 +49,42 @@ public async Task UpdateTurrets(Dictionary<string, TurretData>? turretData = nul
...
+            TurretData turret = kvp.Value;
+
+            string cargoStore =
+                $"{{{{#cargo_store:_table=TurretData|name={turret.Name ?? "N/A"}|size={turret.Size ?? "N/A"}|class={turret.Class ?? "N/A"}|turretsize={turret.TurretSize ?? "N/A"}|dps={turret.Dps:G5}|mass={turret.Mass:G5}|distance={turret.Range:G5}|damage={turret.Damage:G5}|reload={turret.Reload:G5}|beamsize={turret.BeamSize:G5}|override={turret.Override}|maxcycle={turret.MaxCycle}|numbarrels={turret.NumBarrels}|baseaccuracy={turret.BaseAccuracy:G5}|accuracyindex={turret.AccuracyIndex:G5}|rampingstrength={turret.RampingStrength:G5}|speeddenominator={turret.SpeedDenominator:G5}}}}}";
+
+            cargoBuilder.Append(cargoStore);
+        }
+
+        string newTurretDataWikitext = cargoBuilder.ToString();
+        string TurretDataWikitext = await mediaWikiClient.GetArticle("Template:TurretData");
```
Comment (smallketchup82) [inline ketchupbot-framework/TurretUpdater.cs]: I didn't think that we could create a template with all the data and such, then only update that, and set the templates to use the cargo values as their default values.

Bravo, I'm impressed. It might be interesting to see if we can extend this to ship pages as well.

File: ketchupbot-framework/TurretUpdater.cs [hunk 3]
```diff
@@ -47,15 +49,42 @@ public async Task UpdateTurrets(Dictionary<string, TurretData>? turretData = nul
...
+
+        // Including the table declaration on the page currently causes an issue where miraheze never populates the cargo table.
+        // cargoBuilder.Append("{{#cargo_declare:_table=TurretData|name=String|size=String|class=String|turretsize=String|dps=Float|mass=Float|distance=Float|damage=Float|reload=Float|beamsize=Float|override=Boolean|maxcycle=Integer|numbarrels=Integer|baseaccuracy=Float|accuracyindex=Float|rampingstrength=Float|speeddenominator=Float}}");
+
+        foreach (var kvp in turretData)
+        {
+            TurretData turret = kvp.Value;
+
+            string cargoStore =
+                $"{{{{#cargo_store:_table=TurretData|name={turret.Name ?? "N/A"}|size={turret.Size ?? "N/A"}|class={turret.Class ?? "N/A"}|turretsize={turret.TurretSize ?? "N/A"}|dps={turret.Dps:G5}|mass={turret.Mass:G5}|distance={turret.Range:G5}|damage={turret.Damage:G5}|reload={turret.Reload:G5}|beamsize={turret.BeamSize:G5}|override={turret.Override}|maxcycle={turret.MaxCycle}|numbarrels={turret.NumBarrels}|baseaccuracy={turret.BaseAccuracy:G5}|accuracyindex={turret.AccuracyIndex:G5}|rampingstrength={turret.RampingStrength:G5}|speeddenominator={turret.SpeedDenominator:G5}}}}}";
```
Comment (smallketchup82) [inline ketchupbot-framework/TurretUpdater.cs]: Yeah this is a huge mess. I'll refactor this myself

File: ketchupbot-framework/WikiParser.cs [hunk 4]
```diff
@@ -322,16 +322,16 @@ public static MatchCollection ExtractTurretTables(string text)
...
+                "Irregular number of turret tables found. Please ensure that the number of tables stays at 7 or below."),
             _ => turretTables
         };
     }
 
     #region Regexes
 
     [GeneratedRegex(
-        @"{{\s*Ship[ _]Infobox(?:[^{}]|{{[^{}]*}}|{{{[^{}]*}}})+(?:(?!{{(?:[^{}]|{{[^{}]*}}|{{{[^{}]*}}})*)}})",
+        @"{{\s*Ship[ _]Infobox((?:(?>[^{}]*)|\{\{(?:(?>[^{}]*)|\{\{(?>[^{}]*)\}\})*\}})*)}}",
```
Comment (smallketchup82) [inline ketchupbot-framework/WikiParser.cs]: I'll need to take a look at this myself to see if this breaks anything. I do apologize but any changes to the regex are incredibly sensitive so we need to do some heavy testing on this

**Review by smallketchup82 — CHANGES_REQUESTED** [review]

File: ketchupbot-framework/API/ApiManager.cs (hunk 1 above)
Comment (smallketchup82) [reply ketchupbot-framework/API/ApiManager.cs]: Are you using jetbrains rider?

Comment (smallketchup82) [reply ketchupbot-framework/API/ApiManager.cs]: I would HIGHLY recommend using jetbrains rider for c# development, it's free and is more likely to respect any of the hinting/linting rules I've set for the project
